File: HackerRank/Interview_Prep/Arrays/Minimum_Swaps_2/solution.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the minimumSwaps function below.
def minimumSwaps(arr):
    swaps = 0
    di = dict(zip(arr, range(1,len(arr)+1)))
    for i in range(1, len(arr)+1):
        if di[i] != i:
            di[arr[i-1]] = di[i]
            arr[di[i]-1] = arr[i-1]
            swaps += 1
    return swaps

File: HackerRank/Interview_Prep/Arrays/Minimum_Swaps_2/solution.cpp
#include <bits/stdc++.h>

using namespace std;

// Complete the minimumSwaps function below.
int minimumSwaps(vector<int> arr) {
    unordered_map<int,int> dict;
    int swaps = 0;
    int n = arr.size();
    for(int i=0; i<n; ++i)
    {
        dict[arr[i]] = i;
    }
    for(int i=1; i<= n; ++i)
    {
        if(dict[i]+1 != i)
        {
            dict[arr[i-1]] = dict[i];
            arr[dict[i]] = arr[i-1];
            swaps++;
        }
    }
    return swaps;
}